sqlite/sqlite.go

```diff
@@ -81,23 +82,33 @@ func openDB(path string, key string, kdfIterationsNumber int) (*sql.DB, error) {
 		return nil, err
 	}
 
-	// Disable concurrent access as not supported by the driver
-	db.SetMaxOpenConns(1)
-
-	if _, err = db.Exec("PRAGMA foreign_keys=ON"); err != nil {
-		return nil, err
-	}
-	keyString := fmt.Sprintf("PRAGMA key = '%s'", key)
-	if _, err = db.Exec(keyString); err != nil {
-		return nil, errors.New("failed to set key pragma")
-	}
-
-	if kdfIterationsNumber <= 0 {
-		kdfIterationsNumber = sqlite.ReducedKDFIterationsNumber
-	}
-
-	if _, err = db.Exec(fmt.Sprintf("PRAGMA kdf_iter = '%d'", kdfIterationsNumber)); err != nil {
-		return nil, err
+	// Workaround: casting the driver type breaks the database/sql abstraction
+	// and may lead to compatibility issues in the future.
+	// This method is used because the 'database/sql' package does not expose 'ConnectHook',
+	// thereby making it impossible to individually configure each connection.
+	// Consequently, the connections from the pool can't be properly decrypted, making them unusable.
+	sqlcipherDriver, ok := db.Driver().(*sqlcipher.SQLiteDriver)
+	if !ok {
+		return nil, fmt.Errorf("unable to get sqlcipher driver")
+	}
+	sqlcipherDriver.ConnectHook = func(conn *sqlcipher.SQLiteConn) error {
+		if _, err = conn.Exec("PRAGMA foreign_keys=ON", []driver.Value{}); err != nil {
+			return err
+		}
+		keyString := fmt.Sprintf("PRAGMA key = '%s'", key)
+		if _, err = conn.Exec(keyString, []driver.Value{}); err != nil {
+			return errors.New("failed to set key pragma")
+		}
+
+		if kdfIterationsNumber <= 0 {
+			kdfIterationsNumber = sqlite.ReducedKDFIterationsNumber
+		}
+
+		if _, err = conn.Exec(fmt.Sprintf("PRAGMA kdf_iter = '%d'", kdfIterationsNumber), []driver.Value{}); err != nil {
+			return err
+		}
+
+		return nil
```

Comment:
Good catch! I thought setting it once would be sufficient, as once you set PRAGMA journal_mode=WAL;, the setting is stored in the database file itself. I have learned that even though switching the mode to WAL introduces a persistent change to the database, each new connection must still "opt-in" to using the WAL mode. If a connection does not do this, it defaults to the DELETE journal mode. While this connection can still read from and write to the database, it will not take advantage of the benefits of WAL.

Reply:
Updated